db/listings.go

```diff
@@ -0,0 +1,734 @@
+	if err != nil {
+		return nil, err
+	}
+	defer listingRows.Close()
+
+	// Build the listing from the result set.
+	listing := make([]*model.Notification, 0)
+	for listingRows.Next() {
+		var notificationType string
+		var messageText []byte
+		var seen, deleted bool
+		var id, timeCreated string
+
+		// Fetch the data for the current row from the database.
+		err = listingRows.Scan(&notificationType, &seen, &deleted, &messageText, &id, &timeCreated)
+		if err != nil {
+			return nil, err
+		}
+
+		// Unmarshal the message and plug in any values that might have changed.
+		message, err := formatNotification(messageText, notificationType, seen, deleted)
+		if err != nil {
+			return nil, err
+		}
+
+		listing = append(listing, message)
+	}
+
+	return listing, nil
+}
+
+// v2GetNotificationCount determines the total number of notifications that match a notification listing for version 2
+// of the API, not considering paging parameters.
+func v2GetNotificationCount(tx *sql.Tx, params *V2NotificationListingParameters) (int, error) {
+	var total int
+	countQuery, countQueryArgs, err := v2ListNotificationsBaseQuery(params).Column("count(*) AS count").ToSql()
+	if err != nil {
+		return 0, err
+	}
+	err = tx.QueryRow(countQuery, countQueryArgs...).Scan(&total)
+	if err != nil {
+		return 0, err
+	}
+	return total, nil
+}
+
+// runBoundaryIDQueryParams defines the parameters to runBoundaryIDQuery.
+type runBoundaryIDQueryParams struct {
+	Tx                  *sql.Tx
+	QueryBuilder        sq.SelectBuilder
+	WhereFn             func(sq.SelectBuilder, string, *time.Time) (sq.SelectBuilder, error)
+	ComparisonID        string
+	ComparisonTimestamp *time.Time
+	SortOrder           query.SortOrder
+	Offset              uint64
+}
+
+// runBoundaryIDQuery runs a single boundary ID query and returns the result.
+func runBoundaryIDQuery(params runBoundaryIDQueryParams) (string, error) {
+	builder := params.QueryBuilder
+	var err error
+
+	// Add the where clause to the query if necessary.
+	if params.WhereFn != nil {
+		builder, err = params.WhereFn(params.QueryBuilder, params.ComparisonID, params.ComparisonTimestamp)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	// Add the sort order, offset and limit to the query.
+	builder = v2ApplySortOrder(builder, params.SortOrder)
+	builder = builder.Offset(params.Offset).Limit(1)
+
+	// Build the query.
+	query, args, err := builder.ToSql()
+	if err != nil {
+		return "", err
+	}
+
+	// Execute the query.
+	rows, err := params.Tx.Query(query, args...)
+	if err != nil {
+		return "", err
+	}
+	defer rows.Close()
+
+	// Get the ID of the next message if applicable.
+	var id string
+	if rows.Next() {
+		err = rows.Scan(&id)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return id, nil
+}
+
+// v2GetBoundaryIDs obtains the IDs of the messages just beyond the boundaries of the current page.
+func v2GetBoundaryIDs(tx *sql.Tx, params *V2NotificationListingParameters) (string, string, error) {
```

Comment:
So, I admit I'm having trouble following this function, but I'm not quite sure why it's necessary either. I'd expect that instead of having to do any additional queries, we'd look at the first and last items in the listing, and use those for the before and after values to get the pages on either side. If I'm understanding right, which I'm far from confident of, it seems like this function gets the id of the first things _beyond_ the page, which I'd expect to be the wrong values to pass to an additional call (because doing so should exclude those values, e.g. if passing id 345 as 'beforeId' I'd expect it to show notifications with IDs up to 344). I'm sure I'm missing something, anyway -- what's going on here?

Reply:
I originally had the same plan: I was going to return the first and the last IDs in the result set and call it good. I ran into a problem when I was trying to imagine what the UI would do when it encountered the beginning or end of the result set. If I'm always returning the first and last ID on the current page then the UI has know way of knowing whether or not it should be possible to view either prior or subsequent messages without making another API call.

The simplest way to indicate whether or not more results are available in the response body was to use a null value in either the `before_id` or `after_id` fields in the response body when there is no previous or subsequent page, respectively. If I'm returning a null value only when the current page is the first or last page then it made sense to me to make `before_id` and `after_id` in the response body refer to messages that lie just beyond the current page. My thinking would be that if `before_id` is null then there literally is no notification satisfying the other query parameters that comes before the current page.

This does mean that the query parameters, `before-id` and `after-id` might be slightly misnamed, unfortunately. I briefly considered renaming them, but no good names came to mind. I found some examples of APIs that use things such as `filter=id:lte:{some-id}`, but that doesn't accurately describe what this endpoint does either. I eventually decided to leave the query parameter names the same and to provide a better description in the Swagger docs.

With respect to the code itself, would it help to break the cases into separate functions and add more comments? (I could add more comments without breaking each of the cases into a separate function, but I tend not to like large blocks of comments within the body of a function.)

Reply:
Ah, interesting. So, in effect, the parameter being passed in is something like the "earliest ID"/"least ID" or "latest ID" /"greatest ID", and then the before and after IDs returned are... well, the IDs before the current page and after it (where applicable). I guess that there's potentially a sorta-race-condition where a page might return a `before_id` or `after_id` that's not accurate by the time the next request is made (i.e., if a page of 1, 3, 5, 7 says the after_id is 9, but id 8 gets inserted after, the "next page" will skip 8), but it's also certainly true that we need a way to indicate the presence or absence of pages on either side, and that condition is especially rare for something that uses timestamps, like this.

As far as the code itself, I'm not sure. I think part of my confusion is the number of different structures and functions being passed around, and not always in the order I'd expect. Looking at it now, with fresher brain than my late-night review, I think the two biggest things are how much the functions (particularly the `WHERE` clause ones) are getting shuttled around, and the variation in call styles (`builder.IncludedSquirrelFunction(whatever)` vs `v2SomethingOrOther(builder, some, args)`). I'd like it if I could read in order, so that within this function itself, I see a builder getting made, where clauses getting added, then the query getting run, without having to jump between functions to see that path.

From my "sorta like a DBA sometimes, for some reason" perspective, I also worry that we're doing so many queries to get this info, thereby making the DB pull things in and sort them several times.

Not sure if I've said anything helpful here, but I want to stop writing this comment for now :sweat_smile: I'll keep thinking, and maybe something I've said helps in some way.

Reply:
I read over the code again, and I think I've got an idea for refactoring it that will help. I'll probably want to get some other eyes on it just to make sure that the changes actually help to make the code clearer, though.